src/ext/php7/curl_handlers.c

```diff
@@ -0,0 +1,612 @@
+                Z_ADDREF_P(_z3);                  \
+            }                                     \
+        }                                         \
+        ZVAL_COPY_VALUE(z, _z3);                  \
+    } while (0)
+#endif
+
+bool _dd_ext_curl_loaded = false;  // True global -- do not modify after startup
+
+/* "le_curl" is ext/curl's resource type.
+ * "le_curl" is what php_curl.h names this variable
+ */
+ZEND_TLS int le_curl = 0;
+
+/* Cache things we tend to use a few times */
+ZEND_TLS zval _dd_curl_httpheaders = {.u1.type_info = IS_UNDEF};
+ZEND_TLS zval _dd_format_curl_http_headers = {.u1.type_info = IS_UNDEF};
+ZEND_TLS zend_class_entry *_dd_ArrayKVStore_ce = NULL;
+ZEND_TLS zend_class_entry *_dd_GlobalTracer_ce = NULL;
+ZEND_TLS zend_class_entry *_dd_SpanContext_ce = NULL;
+ZEND_TLS zend_function *_dd_ArrayKVStore_putForResource_fe = NULL;
+ZEND_TLS zend_function *_dd_ArrayKVStore_getForResource_fe = NULL;
+ZEND_TLS zend_function *_dd_ArrayKVStore_deleteResource_fe = NULL;
+ZEND_TLS zend_function *_dd_GlobalTracer_get_fe = NULL;
+ZEND_TLS zend_function *_dd_GlobalTracer_inject_fe = NULL;
+ZEND_TLS zend_function *_dd_SpanContext_ctor = NULL;
+ZEND_TLS bool _dd_curl_integration_loaded = false;
+
+// Do not pass things like "parent", "self", "static" -- fully qualified names only!
+static zend_class_entry *_dd_lookup_ce(const char *str, size_t len, bool autoload) {
+    zend_string *name = zend_string_init(str, len, 0);
+#if PHP_VERSION_ID < 70400
+    zend_class_entry *ce = zend_lookup_class_ex(name, NULL, autoload);
+#else
+    zend_class_entry *ce = zend_lookup_class_ex(name, NULL, autoload ? 0 : ZEND_FETCH_CLASS_NO_AUTOLOAD);
+#endif
+    zend_string_release(name);
+    return ce;
+}
+
+/* Returns a zval containing a copy of the string; caller must release.
+ * Makes initialization easier e.g.
+ *     zval putForResource = ddtrace_zval_stringl(ZEND_STRL("putForResource"));
+ *     // don't forget!
+ *     zend_string_release(Z_STR(putForResource));
+ */
+static zval ddtrace_zval_stringl(const char *str, size_t len) {
+    zval zv;
+    ZVAL_STRINGL(&zv, str, len);
+    return zv;
+}
+
+static zval ddtrace_zval_long(zend_long num) {
+    zval zv;
+    ZVAL_LONG(&zv, num);
+    return zv;
+}
+
+static zval ddtrace_zval_null(void) {
+    zval zv;
+    ZVAL_NULL(&zv);
+    return zv;
+}
+
+/**
+ * Calls the method `fname` on the object, or static method on the `ce` if no object.
+ * If you only have 0-2 args, consider zend_call_method_with_{0..2}_params instead
+ *
+ * @param obj May be null
+ * @param ce Must not be null
+ * @param fn_proxy The result of the method lookup is cached here, unless NULL
+ * @param fname Must not be null
+ * @param retval Must not be null
+ * @param argc The number of items in @param argv
+ * @param argv This is a plain C array e.g. `zval args[argc]`
+ * @return
+ */
+static ZEND_RESULT_CODE ddtrace_call_method(zend_object *obj, zend_class_entry *ce, zend_function **fn_proxy,
```

Comment:
I know you're still wrapping things up, but are you planning on extracting these userland-call API's so they can be used in other places?

Reply:
Done! Found in `php7/engine_api.{c,h}`.